src/hotspot/share/jfr/jni/jfrJavaSupport.cpp

```diff
@@ -759,72 +759,58 @@ bool JfrJavaSupport::is_excluded(jobject thread) {
+  const Handle h_mirror(Handle(THREAD, JNIHandles::resolve(clazz)));
+  assert(h_mirror.not_null(), "invariant");
+  fieldDescriptor handler_field_descriptor;
+  const Klass* const field_holder = get_handler_field_descriptor(h_mirror, &handler_field_descriptor, THREAD);
+  if (field_holder == NULL) {
+    // The only reason should be that klass initialization failed.
+    return NULL;
   }
-
-  fieldDescriptor object_field;
-  Klass* g = instance_klass->find_field(
-    vmSymbols::eventHandler_name(),
-    vmSymbols::object_signature(),
-    true, &object_field);
-  if (g != NULL) {
-    oop ret = h_klass_oop->obj_field(object_field.offset());
-    return ret != NULL ? JfrJavaSupport::local_jni_handle(ret, THREAD) : NULL;
-  }
-  assert(f == NULL && g == NULL, "no handler field for class");
-  return NULL;
+  assert(java_lang_Class::as_Klass(h_mirror()) == field_holder, "invariant");
+  oop handler_oop = h_mirror->obj_field(handler_field_descriptor.offset());
+  return handler_oop != NULL ? JfrJavaSupport::local_jni_handle(handler_oop, THREAD) : NULL;
 }
 
 bool JfrJavaSupport::set_handler(jobject clazz, jobject handler, TRAPS) {
   DEBUG_ONLY(JfrJavaSupport::check_java_thread_in_vm(THREAD));
-  const oop klass_oop = JNIHandles::resolve(clazz);
-  assert(klass_oop != NULL, "invariant");
-  const oop handler_oop = JNIHandles::resolve(handler);
-  assert(handler_oop != NULL, "invariant");
-  Klass* klass = java_lang_Class::as_Klass(klass_oop);
   HandleMark hm(THREAD);
-  Handle h_klass_oop(Handle(THREAD, klass->java_mirror()));
-  InstanceKlass* const instance_klass = static_cast<InstanceKlass*>(klass);
-  klass->initialize(CHECK_false);
-
-  fieldDescriptor event_handler_field;
-  Klass* f = instance_klass->find_field(
-    vmSymbols::eventHandler_name(),
-    vmSymbols::jdk_jfr_internal_handlers_EventHandler_signature(),
-    true, &event_handler_field);
-  if (f != NULL) {
-    h_klass_oop->obj_field_put(event_handler_field.offset(), handler_oop);
-    return true;
-  }
-
-  fieldDescriptor object_handler_field;
-  Klass* g = instance_klass->find_field(
-    vmSymbols::eventHandler_name(),
-    vmSymbols::object_signature(),
-    true, &object_handler_field);
-  if (g != NULL) {
-    h_klass_oop->obj_field_put(object_handler_field.offset(), handler_oop);
-    return true;
+  const Handle h_mirror(Handle(THREAD, JNIHandles::resolve(clazz)));
+  assert(h_mirror.not_null(), "invariant");
+  fieldDescriptor handler_field_descriptor;
+  const Klass* const field_holder = get_handler_field_descriptor(h_mirror, &handler_field_descriptor, THREAD);
+  if (field_holder == NULL) {
+    // The only reason should be that klass initialization failed.
+    return false;
   }
-  assert(f == NULL && g == NULL, "no handler field for class");
-  return false;
+  assert(java_lang_Class::as_Klass(h_mirror()) == field_holder, "invariant");
+  const oop handler_oop = JNIHandles::resolve(handler);
+  assert(handler_oop != NULL, "invariant");
```

Comment:
Should this be rather 
```
if (handler_oop == null) {
  return false;
}
```
to have it consistent with the code in `get_handler_field_descriptor()`?

Reply:
Must not be NULL when set.